onprc_ehr/src/org/labkey/onprc_ehr/ONPRC_EHRController.java

```diff
@@ -512,4 +528,108 @@ public void setStartingId(Integer startingId)
+        {
+            super.validateForm(o, errors);
+
+            StudyService ss = StudyService.get();
+            if (ss == null)
+            {
+                errors.reject(ERROR_REQUIRED, "No study");
+                return;
+            }
+
+            int datasetId = ss.getDatasetIdByName(getContainer(), "cases");
+            Dataset dataset = ss.getDataset(getContainer(), datasetId);
+
+            if (dataset == null)
+            {
+                errors.reject(ERROR_REQUIRED, "Cases dataset not found.");
+                return;
+            }
+
+            Domain domain = dataset.getDomain();
+            if (domain == null)
+            {
+                errors.reject(ERROR_REQUIRED, "Cases dataset domain not found.");
+                return;
+            }
+
+            _casesProvisionedName = domain.getStorageTableName();
+            if (_casesProvisionedName == null)
+            {
+                errors.reject(ERROR_REQUIRED, "Cases dataset provisioned name not found.");
+            }
+        }
+
+        @Override
+        public ApiResponse execute(Object form, BindException errors) throws SQLException
+        {
+            Container c = EHRService.get().getEHRStudyContainer(getContainer());
+            TableInfo ti = QueryService.get().getUserSchema(getUser(), c, "study").getTable("cases");
+
+            SQLFragment sql = new SQLFragment("UPDATE c SET c.caseNo = updates.caseNo FROM studydataset." + _casesProvisionedName + " c JOIN (VALUES ");
+
+            logger.info("Starting case number updates.");
+
+            // Stream cases with no case number, sorted by date
+            SimpleFilter filter = SimpleFilter.createContainerFilter(getContainer());
+            filter.addCondition(FieldKey.fromParts("caseNo"), null, CompareType.ISBLANK);
+            ResultSet rs = new TableSelector(ti, filter, new Sort("date")).getResultSet(false, false);
+
+            boolean newBatch = true;
+            int counter = 0;
+
+            while( rs.next() )
+            {
+                if (newBatch) {
+                    newBatch = false;
+                }
+                else {
+                    sql.append(",");
+                }
+                sql.append("(?, ?)");
+                sql.add(rs.getInt("dsrowid"));
+                sql.add(ONPRC_EHRManager.get().getNextCaseNo(c));
+                counter++;
+
+                // Batch boundary
+                if (counter % 1000 == 0)
+                {
+                    newBatch = true;
+                    sql.append(") AS updates(dsrowid, caseNo) ");
+                    sql.append("ON c.dsrowid = updates.dsrowid");
+
+                    new SqlExecutor(ti.getSchema()).execute(sql);
+
+                    logger.info(counter + " total case numbers added.");
+
+                    sql = new SQLFragment("UPDATE c SET c.caseNo = updates.caseNo FROM studydataset." + _casesProvisionedName + " c JOIN (VALUES ");
```

Comment:
Consider making this SQLFragment a separate variable that just gets appended into the new fragment for each batch

Reply:
updated